include/xplugin/xplugin_registry.hpp

```diff
@@ -0,0 +1,147 @@
+/***************************************************************************
+ * Copyright (c) 2023, Dr. Thorsten Beier                                   *
+ *                                                                          *
+ * Distributed under the terms of the BSD 3-Clause License.                 *
+ *                                                                          *
+ * The full license is in the file LICENSE, distributed with this software. *
+ ****************************************************************************/
+#ifndef XPLUGIN_REGISTRY_HPP
+#define XPLUGIN_REGISTRY_HPP
+
+#include <filesystem>
+#include <memory>
+#include <string>
+#include <unordered_map>
+#include <unordered_set>
+
+#include <xplugin/xplugin_config.hpp>
+#include <xplugin/xshared_library.hpp>
+
+namespace xp
+{
+
+template <class FACTORY_BASE>
+class xplugin_registry
+{
+  public:
+    using factory_base_type = FACTORY_BASE;
+
+    using create_plugin_factory_type = factory_base_type *(*)();
+
+    inline xplugin_registry() = default;
+
+    inline std::size_t add_from_directory(const std::filesystem::path &path,
+                                          const std::string &prefix = get_default_library_prefix(),
+                                          const std::string &extension = get_default_library_extension());
+
+    inline std::unique_ptr<factory_base_type> create_factory(const std::string &name);
+
+    inline std::size_t size() const;
+    inline std::unordered_set<std::string> plugin_names();
+
+  private:
+    static inline std::string get_default_library_extension();
+    static inline std::string get_default_library_prefix();
+
+    std::unordered_map<std::string, xshared_library> m_open_libraries;
+    std::unordered_map<std::string, std::filesystem::path> m_locations;
+};
+
+template <class FACTORY_BASE>
+std::size_t xplugin_registry<FACTORY_BASE>::add_from_directory(const std::filesystem::path &path,
+                                                               const std::string &prefix, const std::string &extension)
+{
+    std::size_t n_added = 0;
+    for (const auto &entry : std::filesystem::directory_iterator(path))
+    {
+        if (entry.path().extension() == extension)
+        {
+            std::string name = entry.path().stem().string();
+
+            // remove prefix
+            if (name.substr(0, prefix.size()) == prefix)
```

Comment:
I think we should keep libraries without `prefix` in their path

Reply:
this is not for the path. this is for the name / key in the map.
For a lib at `some/path/libfoobar.so` we  use `fubar` as name/key and `some/path/libfoobar.so` as value path/value